ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningCache.scala

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.interning
+
+import com.daml.dec.DirectExecutionContext
+import com.daml.lf.data.Ref
+import com.daml.logging.LoggingContext
+
+import scala.concurrent.Future
+
+class DomainStringIterators(
+    val parties: Iterator[String],
+    val templateIds: Iterator[String],
+)
+
+trait StringInterningCache extends StringInterning {
+
+  /** Fill StringInterning Cache with entries
+    *
+    * @param domainStringIterators iterators of the new entires
+    * @return If some of the entries were not part of the cache: they will be added, and these will be returned as a interned-id and raw, prefixed string pairs.
+    */
+  def add(domainStringIterators: DomainStringIterators): Iterable[(Int, String)]
+
+  /** Update the StringInterning Cache from persistence
+    *
+    * @param lastStringInterningId this is the "version" of the persistent cache, which from the StringInterningCache can see if it is behind
+    * @return a completion Future: if the cache is behind it will load the missing entries from persistence, and update the cache state
+    */
+  def updateStringInterningCache(lastStringInterningId: Int)(implicit
```

Comment:
Should it not be verboten to call `updateStringInterningCache` after calls to `add`? This is inviting trouble.

Reply:
Good point, how about segregating this interface, and 
* for the read-side only allow `updateStringInterningCache` 
* for the write side's initialization only `updateStringInterningCache`
* and for the write side updating mechanism only `add`